From 489331887bba3a59379e55fcde9055c3e5308fba Mon Sep 17 00:00:00 2001
From: Andy Yan <andy.yan@rock-chips.com>
Date: Thu, 12 Nov 2020 10:36:44 +0800
Subject: [PATCH] drm/rockchip: vop2: Check the max output width for video port

Mark the display mode which request a unsupported output width
as MODE_BAD_HVALUE.

Change-Id: I5d4dbefb8cdd97d54512f6a8e4c9de74d849ea89
Signed-off-by: Andy Yan <andy.yan@rock-chips.com>
---
 drivers/gpu/drm/rockchip/rockchip_drm_vop.h  | 32 ++++++++++----------
 drivers/gpu/drm/rockchip/rockchip_drm_vop2.c |  9 ++----
 drivers/gpu/drm/rockchip/rockchip_vop2_reg.c |  3 ++
 3 files changed, 22 insertions(+), 22 deletions(-)

diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop.h b/drivers/gpu/drm/rockchip/rockchip_drm_vop.h
index d34455b82dad..0e7821812502 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop.h
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop.h
@@ -30,6 +30,16 @@
 #define AFBDC_FMT_RGB565	0x0
 #define AFBDC_FMT_U8U8U8U8	0x5
 #define AFBDC_FMT_U8U8U8	0x4
+#define VOP_FEATURE_OUTPUT_10BIT	BIT(0)
+#define VOP_FEATURE_AFBDC		BIT(1)
+#define VOP_FEATURE_ALPHA_SCALE		BIT(2)
+
+#define WIN_FEATURE_HDR2SDR		BIT(0)
+#define WIN_FEATURE_SDR2HDR		BIT(1)
+#define WIN_FEATURE_PRE_OVERLAY		BIT(2)
+#define WIN_FEATURE_AFBDC		BIT(3)
+#define WIN_FEATURE_CLUSTER_MAIN	BIT(4)
+#define WIN_FEATURE_CLUSTER_SUB		BIT(5)
 
 enum bcsh_out_mode {
 	BCSH_OUT_MODE_BLACK,
@@ -103,6 +113,11 @@ struct vop_csc {
 	uint32_t r2y_offset;
 };
 
+struct vop_rect {
+	int width;
+	int height;
+};
+
 struct vop_ctrl {
 	struct vop_reg version;
 	struct vop_reg standby;
@@ -586,6 +601,7 @@ struct vop2_video_port_data {
 	char id;
 	uint32_t feature;
 	uint64_t soc_id;
+	struct vop_rect max_output;
 	const u8 pre_scan_max_dly[4];
 	const struct vop_intr *intr;
 	const struct vop_hdr_table *hdr_table;
@@ -629,22 +645,6 @@ struct vop_grf_ctrl {
 	struct vop_reg grf_dclk_inv;
 };
 
-#define VOP_FEATURE_OUTPUT_10BIT	BIT(0)
-#define VOP_FEATURE_AFBDC		BIT(1)
-#define VOP_FEATURE_ALPHA_SCALE		BIT(2)
-
-#define WIN_FEATURE_HDR2SDR		BIT(0)
-#define WIN_FEATURE_SDR2HDR		BIT(1)
-#define WIN_FEATURE_PRE_OVERLAY		BIT(2)
-#define WIN_FEATURE_AFBDC		BIT(3)
-#define WIN_FEATURE_CLUSTER_MAIN	BIT(4)
-#define WIN_FEATURE_CLUSTER_SUB		BIT(5)
-
-struct vop_rect {
-	int width;
-	int height;
-};
-
 struct vop_data {
 	const struct vop_reg_data *init_table;
 	unsigned int table_size;
diff --git a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
index b6de49fc86e1..f98544e6ee17 100644
--- a/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
+++ b/drivers/gpu/drm/rockchip/rockchip_drm_vop2.c
@@ -2192,19 +2192,16 @@ vop2_crtc_mode_valid(struct drm_crtc *crtc, const struct drm_display_mode *mode,
 	struct vop2_video_port *vp = to_vop2_video_port(crtc);
 	struct vop2 *vop2 = vp->vop2;
 	const struct vop2_data *vop2_data = vop2->data;
+	const struct vop2_video_port_data *vp_data = &vop2_data->vp[vp->id];
 	int request_clock = mode->clock;
 	int clock;
 
-	if (mode->hdisplay > vop2_data->max_output.width)
+	if (mode->hdisplay > vp_data->max_output.width)
 		return MODE_BAD_HVALUE;
 
-	if ((mode->flags & DRM_MODE_FLAG_INTERLACE) &&
-	    VOP_MAJOR(vop2->version) == 3 &&
-	    VOP_MINOR(vop2->version) <= 2)
-		return MODE_BAD;
-
 	if (mode->flags & DRM_MODE_FLAG_DBLCLK)
 		request_clock *= 2;
+
 	clock = clk_round_rate(vp->dclk, request_clock * 1000) / 1000;
 
 	/*
diff --git a/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c b/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
index 9775ceb9896d..587cf4e6f200 100644
--- a/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
+++ b/drivers/gpu/drm/rockchip/rockchip_vop2_reg.c
@@ -467,6 +467,7 @@ static const struct vop2_video_port_data rk3568_vop_video_ports[] = {
 	 .id = 0,
 	 .soc_id = 0x3568,
 	 .feature = VOP_FEATURE_OUTPUT_10BIT,
+	 .max_output = { 4096, 2304 },
 	 .pre_scan_max_dly = { 40, 49, 33, 42 },
 	 .intr = &rk3568_vp0_intr,
 	 .hdr_table = &rk3568_vop_hdr_table,
@@ -475,6 +476,7 @@ static const struct vop2_video_port_data rk3568_vop_video_ports[] = {
 	{
 	 .id = 1,
 	 .soc_id = 0x3568,
+	 .max_output = { 2048, 1536 },
 	 .pre_scan_max_dly = { 40, 40, 40, 40 },
 	 .intr = &rk3568_vp1_intr,
 	 .regs = &rk3568_vop_vp1_regs,
@@ -482,6 +484,7 @@ static const struct vop2_video_port_data rk3568_vop_video_ports[] = {
 	{
 	 .id = 2,
 	 .soc_id = 0x3568,
+	 .max_output = { 1920, 1080 },
 	 .pre_scan_max_dly = { 40, 40, 40, 40 },
 	 .intr = &rk3568_vp2_intr,
 	 .regs = &rk3568_vop_vp2_regs,
-- 
2.35.3

